File: LeetCode/0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int idx = -1;
        int n = nums.size();
        for (int i = n - 1; i > 0; i--) {
            if (nums[i] > nums[i-1]) {
                idx = i-1;
                break;
            }
        }
        if (idx != -1) {
            int swap_idx = idx;
            for (int i = n - 1; i >= idx; i--) {
                if (nums[i] > nums[idx]) {
                    swap_idx = i;
                    break;
                }
            }
            swap(nums[swap_idx], nums[idx]);
            reverse(nums.begin() + idx + 1, nums.end());
        } else {
            reverse(nums.begin(), nums.end());
        }
    }
};





// class Solution {
// public:
//     void nextPermutation(vector<int>& nums) {
//         next_permutation(nums.begin(), nums.end());
//     }
// };

File: LeetCode/0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length, k = -1, l = -1;
        for (int i = n - 2; i >= 0; i--) 
        {
            if (nums[i] < nums[i + 1]) 
            {
                k = i;
                break;
            }
        }
        if (k == -1) 
        {
            Arrays.sort(nums);
        } 
        else 
        {
            for (int i = n - 1; i > k; i--) 
            {
                if (nums[i] > nums[k]) 
                {
                    l = i;
                    break;
                }
            }
            swap(nums, k, l);
            Arrays.sort(nums, k + 1, n);
        }
    }

    private void swap(int[] nums, int i, int j) 
    {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}